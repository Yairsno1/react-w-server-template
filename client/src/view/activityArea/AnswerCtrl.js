import React, {Component} from 'react';
import PropTypes from 'prop-types';

class AnswerCtrl extends Component {
  constructor(props) {
    super(props);

    this.state = {
      value: NaN,
      submitted: false,
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  //Verify that the input value is always valid number
  handleChange(event) {
    const tmpValue = Number(event.target.value.trim());

    //Todo: handle leading zero
    if (isNaN(tmpValue)) {
      this.setState((currState, props) => ({
        value: isNaN(currState.value) ? '' : currState.value,
      }));
      //const currValue = this.state.value;
      //event.target.value = isNaN(currValue) ? '' : currValue.toString();
    } else {
      this.setState({value: tmpValue});
    }
  }

  handleSubmit(event) {
    event.preventDefault();

    const answer = this.state.value;
    if (answer !== '' && !isNaN(answer)) {
      this.setState({submitted: true});
      this.props.submitAnswer(answer);
    }
  }

  render() {
    const submitBtnClassItems = [
      'w3-button',
      'hp-btn',
      'w3-circle',
      'w3-xlarge',
      'w3-black',
      'w3-text-' + this.props.submitBtnColor,
      'w3-hover-black'];

      const labelClassItems = [
        'w3-padding-16',
        'fa fa-calculator',
        'w3-xxlarge',
        'w3-text-black',
        this.state.submitted ? 'w3-spin' : ''];

    return (
      <form className="w3-container w3-center hp-answer" onSubmit={this.handleSubmit}>
        <label>
          <span className={labelClassItems.join(' ')}></span>
          <input
            type="text"
            name="answer"
            dir="ltr"
            className="w3-xlarge w3-input hp-input w3-border w3-center"
            autoFocus
            autoComplete={"off"}
            value={isNaN(this.state.value) ? '' : this.state.value}
            onChange={this.handleChange}/>
        </label>
        <p className="w3-center">
          <button type="submit" className={submitBtnClassItems.join(' ')}>
            <b>=</b>
          </button>
        </p>
      </form>
    );
  }
}

AnswerCtrl.propTypes = {
  submitBtnColor: PropTypes.string.isRequired,
  submitAnswer: PropTypes.func.isRequired,
};

export default AnswerCtrl;
