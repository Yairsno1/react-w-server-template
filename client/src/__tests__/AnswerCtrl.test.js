import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import AnswerCtrl from '../view/activityArea/AnswerCtrl';

let m_props;

beforeEach( () => {
  m_props = {
    submitBtnColor: 'red',
    submitAnswer: jest.fn(),
  };
  m_props.submitAnswer.mockReset();
});

it('AnswerCtrl_rendersWithoutCrashing', () => {
  mount(<AnswerCtrl {...m_props}/>);
});

it('AnswerCtrl_snapshot', () => {
  const answerCtrlComp = renderer.create(<AnswerCtrl {...m_props}/>);

  let actualAnswerCtrlComp = answerCtrlComp.toJSON();

  expect(actualAnswerCtrlComp).toMatchSnapshot();
});

it('AnswerCtrl_submitBtnTextColor', () => {
  const expectedTextColor = 'w3-text-' + m_props.submitBtnColor;

  const answerCtrlComp = shallow(<AnswerCtrl {...m_props}/>);

  const actualBtn = answerCtrlComp.find('button');

  expect(actualBtn).toHaveClassName(expectedTextColor);
});

it('AnswerCtrl_notSpinningOnMount', () => {
  const expectedSpinClass = 'w3-spin'

  const answerCtrlComp = shallow(<AnswerCtrl {...m_props}/>);

  const actualSppiner = answerCtrlComp.find('span');

  expect(actualSppiner).not.toHaveClassName(expectedSpinClass);
});

it('AnswerCtrl_emptyInputOnMount', () => {
  const expectedVal = '';

  const answerCtrlComp = mount(<AnswerCtrl {...m_props}/>);

  const actualVal = answerCtrlComp.find('input').instance().value;

  expect(actualVal).toEqual(expectedVal);
});

it('AnswerCtrl_invalidInput', () => {
  const expectedVal = '';

  const answerCtrlComp = mount(<AnswerCtrl {...m_props}/>);

  const input = answerCtrlComp.find('input');
  input.simulate('change', {
    target: {
      value: 'This is not a valid number',
    }
   });
  const actualVal = input.instance().value;

  expect(actualVal).toEqual(expectedVal);
});

it('AnswerCtrl_validInput', () => {
  const expectedVal = 100;

  const answerCtrlComp = mount(<AnswerCtrl {...m_props}/>);

  const input = answerCtrlComp.find('input');
  input.simulate('change', {
    target: {
      value: expectedVal.toString(),
    }
  });
  const actualVal = Number(input.instance().value);

  expect(actualVal).toEqual(expectedVal);
});

it('AnswerCtrl_submit_validInput', () => {
  const answer = 100;

  const answerCtrlComp = mount(<AnswerCtrl {...m_props}/>);

  const input = answerCtrlComp.find('input');
  input.simulate('change', {
    target: {
      value: answer.toString(),
    }
  });
  answerCtrlComp.find('form').simulate('submit');

  expect(m_props.submitAnswer).toHaveBeenCalled();
});

it('AnswerCtrl_submit_spinner_validInput', () => {
  const expectedSpinClass = 'w3-spin';
  const answer = 100;

  const answerCtrlComp = mount(<AnswerCtrl {...m_props}/>);

  const input = answerCtrlComp.find('input');
  input.simulate('change', {
    target: {
      value: answer.toString(),
    }
  });
  answerCtrlComp.find('form').simulate('submit');

  const actualSppiner = answerCtrlComp.find('span');

  expect(actualSppiner).toHaveClassName(expectedSpinClass);
});

it('AnswerCtrl_submit_invalidInput', () => {
  const answer = '10z';

  const answerCtrlComp = mount(<AnswerCtrl {...m_props}/>);

  const input = answerCtrlComp.find('input');
  input.simulate('change', {
    target: {
      value: answer.toString(),
    }
  });
  answerCtrlComp.find('form').simulate('submit');

  expect(m_props.submitAnswer).not.toHaveBeenCalled();
});

it('AnswerCtrl_submit_spinner_invalidInput', () => {
  const expectedSpinClass = 'w3-spin';
  const answer = '10z';

  const answerCtrlComp = mount(<AnswerCtrl {...m_props}/>);

  const input = answerCtrlComp.find('input');
  input.simulate('change', {
    target: {
      value: answer.toString(),
    }
  });
  answerCtrlComp.find('form').simulate('submit');

  const actualSppiner = answerCtrlComp.find('span');

  expect(actualSppiner).not.toHaveClassName(expectedSpinClass);
});

it('AnswerCtrl_submit_submittedAnswer', () => {
  const expectedAnswer = 100;

  const answerCtrlComp = mount(<AnswerCtrl {...m_props}/>);

  const input = answerCtrlComp.find('input');
  input.simulate('change', {
    target: {
      value: expectedAnswer.toString(),
    }
  });
  answerCtrlComp.find('form').simulate('submit');

  expect(m_props.submitAnswer).lastCalledWith(expectedAnswer); // â‰¡ expect(m_props.submitAnswer.mock.calls[0][0]).toEqual(expectedAnswer);
});

