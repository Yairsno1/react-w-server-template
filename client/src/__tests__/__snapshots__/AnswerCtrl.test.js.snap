// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnswerCtrl_snapshot 1`] = `
<form
  className="w3-container w3-center hp-answer"
  onSubmit={[Function]}
>
  <label>
    <span
      className="w3-padding-16 fa fa-calculator w3-xxlarge w3-text-black "
    />
    <input
      autoComplete="off"
      autoFocus={true}
      className="w3-xlarge w3-input hp-input w3-border w3-center"
      dir="ltr"
      name="answer"
      onChange={[Function]}
      type="text"
      value=""
    />
  </label>
  <p
    className="w3-center"
  >
    <button
      className="w3-button hp-btn w3-circle w3-xlarge w3-black w3-text-red w3-hover-black"
      type="submit"
    >
      <b>
        =
      </b>
    </button>
  </p>
</form>
`;
